wordpress-mobile/WordPress-iOS
# Increase performance of Total Likes cell attributed guide text creation

Partially fixes #20849

## Description

I was looking at the particular places slowing down the loading of Stats that [could cause double-navigation issues](https://github.com/wordpress-mobile/WordPress-iOS/issues/20849#issuecomment-1639995735) leading to a crash. At least for the Insights tab, it was enough for it to contain `Total Likes` card with `NSAttributedString` generated from HTML via `makeTotalInsightsGuideText` method. Even just having this card alone with the attributed text I can reproduce double navigation almost every time. I couldn't reproduce the issue with any other card.

Due to Stats inefficient implementation, table view gets multiple reloads when opening Stats. In my tests on iPhone 14 Pro, loading of data for `Total Likes` cell with the attributed text could total up to `1.2sec`, compared to `0.003sec` of total loading data time if the attributed text didn't exist.

This is how time profiler looks after spending 8 seconds in Stats, simply switching once between Insights/Days/Weeks/Months/Years. `NSAttributedString` initializer was the most consuming activity, and it was only done for total likes cell:

<img width="1154" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/88669f89-bb1e-4f93-8303-da2f4958e6da">

It turns out this is a known issue:
- https://stackoverflow.com/questions/31852655/very-slow-html-rendering-in-nsattributedstring 
- https://www.robpeck.com/2015/04/nshtmltextdocumenttype-is-slow/

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/4475bf58-b123-4f39-a9b2-04177f590220

## Solution

Replace HTML-driven solution with a manual attributed string building solution:
- Detect a range of the first placeholder of the localized string (Post title)
- Detect a range of the second placeholder of the localized string (Likes count)
- Apply text attributes

In my tests total likes cell row data is loaded 35 times faster, and I wasn't able to reproduce double-navigation issues anymore.

### Other improvements

Although the issue was driven by the slow html-driven attributed string building, the way Stats Insights table is built also contributes to the issue. Initially, the table view is reloaded 12 times even just with a single total likes row. Implementing data caching, and only reloading parts of the UI that require updates would also make such inefficient methods less visible. 

### To test

1. Open Jeptack
2. Login and choose a site that has a post written
3. Open Stats (Make sure the Total likes card is added)
4. Confirm guide text appears as before

To test double-navigation I found it easier to reproduce on the simulator by pressing "Stats" quickly multiple times with the mouse. You can try it on the trunk and confirm you can reproduce it before trying out the functionality on this branch.

## Regression Notes
1. Potential unintended areas of impact

Guide text not working anymore

6. What I did to test those areas of impact (or what existing automated tests I relied on)

Manual testing

7. What automated tests I added (or what prevented me from doing so)

None

PR submission checklist:


UI Changes testing checklist:
